File: java/ShipDescription.java
import com.sun.jna.Pointer;

///Base class for all variables needed to define a ship
abstract class ShipDescription
{
  Pointer ptr;
  int ID;
  int iteration;
  public ShipDescription(Pointer p)
  {
    ptr = p;
    ID = Client.INSTANCE.shipDescriptionGetId(ptr);
    iteration = BaseAI.iteration;
  }
  abstract boolean validify();

    //commands


    //getters

  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetId(ptr);
  }
  ///The ship type
  public String getType()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetType(ptr);
  }
  ///The amount of energy required to warp in this type of ship
  public int getCost()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetCost(ptr);
  }
  ///The radius of the ship
  public int getRadius()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetRadius(ptr);
  }
  ///The range of attacks for this ship, given as the maximum distance from the center of this ship to the edge of the target
  public int getRange()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetRange(ptr);
  }
  ///The strength of attacks for this ship
  public int getDamage()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetDamage(ptr);
  }
  ///The amount of damage done when this ship self destructs
  public int getSelfDestructDamage()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetSelfDestructDamage(ptr);
  }
  ///The largest possible movement for this ship
  public int getMaxMovement()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetMaxMovement(ptr);
  }
  ///The max number of attacks for this ship
  public int getMaxAttacks()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetMaxAttacks(ptr);
  }
  ///The max health possible for the ship
  public int getMaxHealth()
  {
    validify();
    return Client.INSTANCE.shipDescriptionGetMaxHealth(ptr);
  }

}

File: cpp/ShipDescription.cpp
// -*-c++-*-

#include "ShipDescription.h"
#include "game.h"


ShipDescription::ShipDescription(_ShipDescription* pointer)
{
    ptr = (void*) pointer;
}

int ShipDescription::id()
{
  return ((_ShipDescription*)ptr)->id;
}

char* ShipDescription::type()
{
  return ((_ShipDescription*)ptr)->type;
}

int ShipDescription::cost()
{
  return ((_ShipDescription*)ptr)->cost;
}

int ShipDescription::radius()
{
  return ((_ShipDescription*)ptr)->radius;
}

int ShipDescription::range()
{
  return ((_ShipDescription*)ptr)->range;
}

int ShipDescription::damage()
{
  return ((_ShipDescription*)ptr)->damage;
}

int ShipDescription::selfDestructDamage()
{
  return ((_ShipDescription*)ptr)->selfDestructDamage;
}

int ShipDescription::maxMovement()
{
  return ((_ShipDescription*)ptr)->maxMovement;
}

int ShipDescription::maxAttacks()
{
  return ((_ShipDescription*)ptr)->maxAttacks;
}

int ShipDescription::maxHealth()
{
  return ((_ShipDescription*)ptr)->maxHealth;
}




std::ostream& operator<<(std::ostream& stream,ShipDescription ob)
{
  stream << "id: " << ((_ShipDescription*)ob.ptr)->id  <<'\n';
  stream << "type: " << ((_ShipDescription*)ob.ptr)->type  <<'\n';
  stream << "cost: " << ((_ShipDescription*)ob.ptr)->cost  <<'\n';
  stream << "radius: " << ((_ShipDescription*)ob.ptr)->radius  <<'\n';
  stream << "range: " << ((_ShipDescription*)ob.ptr)->range  <<'\n';
  stream << "damage: " << ((_ShipDescription*)ob.ptr)->damage  <<'\n';
  stream << "selfDestructDamage: " << ((_ShipDescription*)ob.ptr)->selfDestructDamage  <<'\n';
  stream << "maxMovement: " << ((_ShipDescription*)ob.ptr)->maxMovement  <<'\n';
  stream << "maxAttacks: " << ((_ShipDescription*)ob.ptr)->maxAttacks  <<'\n';
  stream << "maxHealth: " << ((_ShipDescription*)ob.ptr)->maxHealth  <<'\n';
  return stream;
}
